### PR TITLE
[ticket/16604] Display warning messages on avatar upload setting issues

### PR DESCRIPTION
**Pull request by rxu**
Checklist:

- [x] Correct branch: master for new features; 3.3.x & 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html), [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

<a href="https://tracker.phpbb.com/browse/PHPBB3-16604">PHPBB3-16604</a>.

### CODE REVIEW

File: phpBB/includes/acp/acp_board.php [hunk 1]
```diff
@@ -510,6 +510,25 @@ function main($id, $mode)
...
+			// If avatar uploading is enabled but the path setting is empty,
+			// config variable validation is bypassed. Catch the case here
+			if (!$cfg_array['avatar_path'])
+			{
+				$error[] = $language->lang('AVATAR_NO_UPLOAD_PATH');
+			}
+			// Not existing or writable path will be caught on submit by config vars validation
+			// Display the warning if the directory was changed on the server afterwards
+			else if (!$submit &&
+				(!$filesystem->exists($phpbb_root_path . $cfg_array['avatar_path']) || !$filesystem->is_writable($phpbb_root_path . $cfg_array['avatar_path'])))
```
Comment (3D-I) [inline phpBB/includes/acp/acp_board.php:526]: Something more readable here?

Comment (rxu) [reply phpBB/includes/acp/acp_board.php:526]: Suugestions? :)

Comment (3D-I) [reply phpBB/includes/acp/acp_board.php:526]: See above, just a suggestion in fact.

Comment (3D-I) [reply phpBB/includes/acp/acp_board.php:526]: Or

```
			else if (!$submit &&
				(
					!$filesystem->exists($phpbb_root_path . $cfg_array['avatar_path'])
					||
					!$filesystem->is_writable($phpbb_root_path . $cfg_array['avatar_path'])
				)
			)

```

**Comment by 3D-I** [comment]
Nice. :+1:

File: phpBB/language/en/common.php [hunk 2]
```diff
@@ -110,6 +110,8 @@
 	'AVATAR_NOT_UPLOADED'			=> 'Avatar could not be uploaded.',
 	'AVATAR_NO_TEMP_DIR'			=> 'Temporary folder could not be found or is not writable.',
 	'AVATAR_NO_SIZE'				=> 'The width or height of the linked avatar could not be determined. Please enter them manually.',
+	'AVATAR_NO_UPLOAD_DIR'			=> 'Avatar storage path does not exist or is not writable.',
+	'AVATAR_NO_UPLOAD_PATH'			=> 'Avatar uploading is enabled but avatar storage path is not set.',
```
Comment (3D-I) [inline phpBB/language/en/common.php:114]: * No storage path available, either it does not exists or is not writeable.
* Avatar storage path not set, although avatar upload is enabled.

Just a suggestion, hopefully some mother tongue will chime on this.

**Comment by rxu** [comment]
I\m not sure how toi fix `chmod 444` for Windows tests, probably just by bypassing it for appveyor.

File: phpBB/includes/acp/acp_board.php [hunk 3]
```diff
@@ -510,6 +510,30 @@ function main($id, $mode)
 			}
 		}
 
+		if ($mode == 'avatar' && $cfg_array['allow_avatar_upload'])
+		{
+			$filesystem = $phpbb_container->get('filesystem');
```
Comment (marc1706) [inline phpBB/includes/acp/acp_board.php:515]: Move this to inside the else branch. It's not needed if the avatar path is empty.

File: phpBB/language/en/common.php (hunk 2 above)
Comment (marc1706) [reply phpBB/language/en/common.php:114]: I think the currently proposed wording is fine and in line with already existing language variables.

File: phpBB/includes/acp/acp_board.php [hunk 4]
```diff
@@ -510,6 +510,29 @@ function main($id, $mode)
...
 
+		if ($mode == 'avatar' && $cfg_array['allow_avatar_upload'])
+		{
+			// If avatar uploading is enabled but the path setting is empty,
+			// config variable validation is bypassed. Catch the case here
+			if (!$cfg_array['avatar_path'])
+			{
+				$error[] = $language->lang('AVATAR_NO_UPLOAD_PATH');
+			}
+			else
```
Comment (marc1706) [inline phpBB/includes/acp/acp_board.php:521]: Change this to `else if (!$submit)`. The filesystem check is only needed then and the check for `$submit` can be removed from the next `if()`

**Review by marc1706 — APPROVED** [review]

**Comment by rxu** [comment]
Regrettably, this seems to be not compatible with 4.x/master storage system as some parts of code were moved to `includes/acp_storage.php`.

**Comment by marc1706** [comment]
Yep, will need to fix master with that regard. :)

**Pull request merged**